src/client/activeobjectmgr.cpp

```diff
@@ -25,28 +25,33 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 namespace client
 {
 
-void ActiveObjectMgr::clear()
+ActiveObjectMgr::~ActiveObjectMgr()
 {
-	// delete active objects
-	for (auto &active_object : m_active_objects) {
-		delete active_object.second;
-		// Object must be marked as gone when children try to detach
-		active_object.second = nullptr;
+	if (!m_active_objects.empty()) {
+		warningstream << "client::ActiveObjectMgr::~ActiveObjectMgr(): not cleared."
+				<< std::endl;
+		clear();
 	}
-	m_active_objects.clear();
 }
 
 void ActiveObjectMgr::step(
 		float dtime, const std::function<void(ClientActiveObject *)> &f)
 {
 	g_profiler->avg("ActiveObjectMgr: CAO count [#]", m_active_objects.size());
-	for (auto &ao_it : m_active_objects) {
-		f(ao_it.second);
+
+	// Same as in server activeobjectmgr.
+	std::vector<u16> ids = getAllIds();
```

Comment:
with many entites you will copy a huge memory block. Use ref as parameter to write result on

Reply:
Do you mean something like `void getAllIds(std::vector<u16> &ret)`? That's not more efficient. Function calls are prvalue expressions (see value categories), and there's copy elision. So, there should only be move constructor calls, no copying.